### PR TITLE
feat/add-lz-share-mover

### PR DESCRIPTION
**Pull request by Rhovian**
- adds `lz-share-mover` program
- modifies `boring-vault-svm` to include three new instructions, `MintShares`, `BurnShares` and `SetShareMover`
- addresses clippy issues across programs
- creates a generic crate for common functionality

### CODE REVIEW

File: programs/common/src/message.rs [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+    buffer.extend_from_slice(&msg.recipient);
+
+    // Amount (16 bytes, big endian)
+    buffer.extend_from_slice(&msg.amount.to_be_bytes());
+
+    buffer
+}
+
+/// Decode bytes into a ShareBridgeMessage
+pub fn decode_message(data: &[u8]) -> Result<ShareBridgeMessage> {
```
Comment (corsur) [inline programs/common/src/message.rs:96]: feels like we should be able to do this automatically with an account definition, right? same for encode? are we trying to avoid accounts since it's just messages?

File: programs/common/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+use anchor_lang::error_code;
+
+#[error_code]
+pub enum MathError {
```
Comment (corsur) [inline programs/common/src/error.rs:4]: does this one not already exist?

File: programs/common/src/rate_limit.rs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+            last_updated: 0,
+            limit: 0,  // A limit of 0 means rate limiting is disabled
+            window: 0, // A window of 0 means rate limiting is disabled
+        }
+    }
+}
+
+impl RateLimitState {
+    /// Checks if the amount can be sent/received within rate limits and updates state if allowed
+    pub fn check_and_consume(&mut self, amount: u64, current_timestamp: i64) -> Result<()> {
```
Comment (corsur) [inline programs/common/src/rate_limit.rs:30]: do we not need to make sure this isn't callable from outside here?

File: programs/layer-zero-share-mover/src/processor/lz_receive.rs [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+use common::message::decode_message;
+
+// min accounts len for clear cpi, found here:
+// https://github.com/LayerZero-Labs/LayerZero-v2/blob/main/packages/layerzero-v2/solana/programs/programs/endpoint/src/instructions/oapp/clear.rs
+pub const CLEAR_MIN_ACCOUNTS_LEN: usize = 5;
+pub const MINT_ACCOUNTS_LEN: usize = 5;
+
+// Add discriminators as constants for better maintainability
+pub const CLEAR_DISCRIMINATOR: [u8; 8] = [250, 39, 28, 213, 123, 163, 133, 5];
+pub const MINT_SHARES_DISCRIMINATOR: [u8; 8] = [24, 196, 132, 0, 183, 158, 216, 142];
```
Comment (corsur) [inline programs/layer-zero-share-mover/src/processor/lz_receive.rs:25]: probably want to centralize this somewhere.

**Review by corsur — COMMENTED** [review]
got through half of this, will review the other half later.

personally i probably would have focused on basic functionality with layer zero of just encode and decode first.

then share moving after.

then rate limiting and other after.

i feel like writing it all at once is a recipe for more work since none of the code is really settled.

File: programs/layer-zero-share-mover/src/processor/deploy.rs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+};
+use anchor_lang::{
+    prelude::*,
+    solana_program::{instruction::Instruction, program::invoke_signed},
+};
+use anchor_spl::token_interface::Mint;
+use common::rate_limit::RateLimitState;
+use std::mem::size_of;
+
+const OAPP_REGISTER_DISCRIMINATOR: [u8; 8] = [129, 89, 71, 68, 11, 82, 210, 125];
```
Comment (corsur) [inline programs/layer-zero-share-mover/src/processor/deploy.rs:17]: is this something we should have their idl for?

File: programs/layer-zero-share-mover/src/processor/send.rs [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+        .map_err(|e| {
+            msg!("LayerZero send CPI failed: {}", e);
+            anchor_lang::error::Error::from(e)
+        })?;
+
+        Ok(())
+    }
+}
+
+pub fn send<'info>(
```
Comment (corsur) [inline programs/layer-zero-share-mover/src/processor/send.rs:256]: hey i've been seeing this in some of these functions. i really think you need to try to fit the functions into the CEI pattern: Checks, Effects, Interactions.

for instance in this function is see checks spread in a couple of places.

File: programs/layer-zero-share-mover/src/processor/send.rs [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+    token_interface::{Mint, TokenAccount},
+};
+use boring_vault_svm::{BoringVault, BASE_SEED_BORING_VAULT_STATE};
+use common::{
+    error::ShareBridgeCodecError,
+    message::{encode_message, ShareBridgeMessage},
+};
+
+const SEND_DISCRIMINATOR: [u8; 8] = [102, 251, 20, 187, 65, 75, 12, 69];
+const BURN_SHARES_DISCRIMINATOR: [u8; 8] = [98, 168, 88, 31, 217, 221, 191, 214];
```
Comment (corsur) [inline programs/layer-zero-share-mover/src/processor/send.rs:25]: idl?

File: programs/layer-zero-share-mover/src/processor/set_executor_program.rs [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+        seeds = [SHARE_MOVER_SEED, share_mover.mint.as_ref()],
+        bump = share_mover.bump,
+    )]
+    pub share_mover: Account<'info, ShareMover>,
+}
+
+pub fn set_executor_program(ctx: Context<SetExecutorProgram>, executor: Pubkey) -> Result<()> {
+    ctx.accounts.share_mover.executor_program = executor;
+    Ok(())
+}
```
Comment (corsur) [inline programs/layer-zero-share-mover/src/processor/set_executor_program.rs:22]: is this the same as endpoint program?

File: programs/layer-zero-share-mover/src/processor/transfer_authority.rs [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+        seeds = [SHARE_MOVER_SEED, share_mover.mint.as_ref()],
+        bump = share_mover.bump,
+    )]
+    pub share_mover: Account<'info, ShareMover>,
+}
+
+pub fn transfer_authority(ctx: Context<TransferAuthority>, new_admin: Pubkey) -> Result<()> {
+    ctx.accounts.share_mover.admin = new_admin;
+    msg!("ShareMover admin updated to: {}", new_admin);
+    Ok(())
```
[the remark this answers is not in the document]
Comment (corsur) [reply programs/layer-zero-share-mover/src/processor/transfer_authority.rs:22]: This is the typical pattern we have used elsewhere, i would recommend following it.

File: programs/layer-zero-share-mover/src/state/lz.rs [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+// https://github.com/LayerZero-Labs/devtools/blob/main/examples/oapp-solana/programs/my_oapp/src/state/peer_config.rs#L7
+#[account]
+pub struct PeerConfig {
+    pub peer_address: [u8; 32],
+    pub enforced_options: EnforcedOptions,
+    pub bump: u8,
+}
+
+pub const ENFORCED_OPTIONS_SEND_MAX_LEN: usize = 512;
+pub const ENFORCED_OPTIONS_SEND_AND_CALL_MAX_LEN: usize = 1024;
```
Comment (corsur) [inline programs/layer-zero-share-mover/src/state/lz.rs:26]: centralize these?

File: programs/layer-zero-share-mover/src/processor/deploy.rs (hunk 5 above)
Comment (Rhovian) [reply programs/layer-zero-share-mover/src/processor/deploy.rs:17]: unfortunately they used anchor 0.29 so i don't think the idl approach would work

File: programs/layer-zero-share-mover/src/processor/set_executor_program.rs (hunk 8 above)
Comment (Rhovian) [reply programs/layer-zero-share-mover/src/processor/set_executor_program.rs:22]: no, different programs

File: programs/layer-zero-share-mover/src/processor/lz_receive.rs (hunk 4 above)
Comment (Rhovian) [reply programs/layer-zero-share-mover/src/processor/lz_receive.rs:25]: I like having these where they are used, the files are manageable atm. If they are used in multiple spots then yeah i'll centralize

File: programs/boring-vault-svm/src/lib.rs [hunk 11]
```diff
@@ -1185,6 +1191,74 @@ pub mod boring_vault_svm {
...
+            &ctx.accounts.vault.config.vault_id.to_le_bytes()[..],
+            &[ctx.bumps.vault],
+        ];
+        let signer_seeds = &[&vault_seeds[..]];
+
+        let mint_to_ctx = CpiContext::new_with_signer(
+            ctx.accounts.token_program.to_account_info(),
+            MintTo {
+                mint: ctx.accounts.share_mint.to_account_info(),
+                to: ctx.accounts.recipient_token_account.to_account_info(),
```
[the remark this answers is not in the document]
Comment (Rhovian) [reply programs/boring-vault-svm/src/lib.rs:1236]: `recipient` is the owner of the ata we use here `recipient_token_account`

**Comment by Rhovian** [comment]
@greptileai

File: programs/endpoint-mock/src/instructions/oapp/register_oapp.rs [hunk 12]
```diff
@@ -0,0 +1,32 @@
+use crate::*;
+
+#[derive(Accounts)]
+#[instruction(params: RegisterOAppParams)]
+pub struct RegisterOApp<'info> {
+    #[account(mut)]
+    pub payer: Signer<'info>,
+    /// The PDA of the OApp
+    pub oapp: Signer<'info>,
```
[the remark this answers is not in the document]
Comment (Rhovian) [reply programs/endpoint-mock/src/instructions/oapp/register_oapp.rs:9]: mock taken directly from LZ

File: programs/boring-vault-svm/src/utils/math.rs [hunk 13]
```diff
@@ -351,18 +343,18 @@ mod tests {
     #[test]
     fn test_math_error_boundaries() {
         // Test that functions handle edge cases gracefully
-        
+
         // Very large premium should not panic
         let large_shares = Decimal::from(u64::MAX);
         let result = apply_share_premium(large_shares, 9999); // 99.99%
         assert!(result.is_ok() || result.is_err()); // Should either work or fail gracefully
```
[the remark this answers is not in the document]
Comment (Rhovian) [reply programs/boring-vault-svm/src/utils/math.rs:350]: @corsur bad test, will likely fix in this PR

File: programs/layer-zero-share-mover/src/state/share_mover.rs [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+    Sui,
+}
+
+impl Default for PeerChain {
+    fn default() -> Self {
+        Self::Unknown
+    }
+}
+
+// TODO: review immutability
```
[the remark this answers is not in the document]
Comment (Rhovian) [reply programs/layer-zero-share-mover/src/state/share_mover.rs:23]: I think its good as is, below i will explain reasoning:

```rust
#[derive(InitSpace)]
#[account]
pub struct ShareMover {
    pub admin: Pubkey,
    pub endpoint_program: Pubkey,
    pub executor_program: Pubkey,
    // immutable after deployment
    pub boring_vault_program: Pubkey,
    // immutable after deployment
    pub vault: Pubkey,
    // immutable after deployment
    pub mint: Pubkey,
    pub is_paused: bool,
    // immutable after deployment
    pub peer_decimals: u8,
    pub bump: u8,
    pub allow_from: bool,
    pub allow_to: bool,
    pub outbound_rate_limit: RateLimitState,
    pub inbound_rate_limit: RateLimitState,
    // immutable after deployment
    pub peer_chain: PeerChain,
}
```

the `share_mover` is derived by a unique mint, which is in turn associated with a specific vault, in a specific program, meaning if these change at all, we should force the user to deploy a new share mover.

Same thing with the peer chain and peer decimals, on EVM these are tied to specific contracts

**Review by corsur — APPROVED** [review]

File: crates/common/src/message.rs [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+
+impl ShareBridgeMessage {
+    pub fn new(recipient: [u8; 32], amount: u128) -> Self {
+        Self { recipient, amount }
+    }
+
+    /// Helper to convert amount between different decimal representations
+    /// Returns None if:
+    /// - Arithmetic overflow would occur
+    /// - The resulting amount would be zero (dust protection)
```
[the remark this answers is not in the document]
Comment (Rhovian) [reply crates/common/src/message.rs:39]: they are handled in `send`

**Review by corsur — APPROVED** [review]
ship it
